Pass msve-vector-bits to Flang to embed in .ll files / release_13x

pwisniewskimobica opened the pull request:
It is duplicate of https://github.com/flang-compiler/classic-flang-llvm-project/pull/69 but add more variants to test file.

Depends on #67
Depends on #68

Pass LLVM msve_vector_bits (simple string) to flang to embed in generated .ll files. For normal compilation this won't make much difference as the attributes are passed to clang after flang2 and can be applied there but this is crucial to enable LTO with flang as clang will not apply the attributes when building the flang2 output. libLTO will need to read these out of the object files to apply them.
For proper building classic llvm tests i had to add new option to build-llvm-project.sh to  decide which additional subprojects should be build. By default there are open mp and clang but for test purposes (this ticket) only clang is selected.

abrahamtovarmob committed:
LLVM release_13x CI changes

Changes to LLVM release_13x to be able to be build by github

Conflicts:
	.github/workflows/Ubuntu20Dockerfile

RichBarton-Arm committed:
[driver] Pass LLVM target_features to flang

Pass LLVM target features (-mattr strings) to flang to embed in generated
.ll files. For normal compilation this won't make much difference as the
attributes are passed to clang after flang2 and can be applied there but this
is crucial to enable LTO with flang as clang will not apply the attributes
when building the flang2 output. libLTO will need to read these out of the
object files to apply them.

Signed-off-by: Richard Barton <richard.barton@arm.com>

RichBarton-Arm committed:
[driver] Add infrastructure for testing the flang-driver in lit

Add %flang as a tool substitution available in lit tests. This apes the way
%clang is defined and adds a $FLANG override in a similar vein.

To avoid this being dead code, add a single test to check the flang driver is
reporting the correct phases when running under various phase control options.

Signed-off-by: Richard Barton <richard.barton@arm.com>

RichBarton-Arm committed:
[flang-driver] Uniquify the target_features string

Use existing clang support for uniquifying target feature lists to pass
a unique list through to flang.

abrahamtovarmob committed:
[driver] Pass LLVM msve-vector-bits to flang

Pass LLVM msve_vector_bits (simple string) to flang to embed in
generated .ll files. For normal compilation this won't make much
difference as the attributes are passed to clang after flang2 and
can be applied there but this is crucial to enable LTO with flang
as clang will not apply the attributes when building the flang2
output. libLTO will need to read these out of the object files
to apply them.

pwisniewskimobica commented:
@bryanpkc @kiranchandramohan @michalpasztamobica @shivaramaarao 
Guys please take a look and mark as reviewed.
Thanks

michalpasztamobica reviewed (approved):
From the CI perspective looks good.

bryanpkc commented on clang/test/Driver/emit-flang-attrs.f90:
```diff
@@ -0,0 +1,10 @@
+// REQUIRES: aarch64-registered-target
```
This test should be moved into `clang/test/Driver/flang/`. Also, the other existing flang driver tests should be updated to use `%flang` instead of `%clang --driver-mode=flang`. On the other hand, this seems to conflict with LLVM Flang; have you considered reducing potential conflicts with upstream? We were hoping to support building both LLVM Flang and Classic Flang from the same code base.

pawosm-arm replied:
The moving of this file to flang directory is done by other commit as this commit is authored by someone else and is contributed as-is.

bryanpkc replied:
OK thanks.

bryanpkc reviewed (changes requested):
Please remove the merge commit from this branch.

pwisniewskimobica commented:
> 

done

pwisniewskimobica commented:
@bryanpkc  
Just a gentle ping that Your comments were applied.

bryanpkc commented on clang/lib/Driver/ToolChains/ClassicFlang.cpp:
```diff
@@ -989,6 +989,20 @@ void ClassicFlang::ConstructJob(Compilation &C, const JobAction &JA,
+    }
+    else
+    // Handle the unsupported values passed to msve-vector-bits.
+    D.Diag(diag::warn_drv_clang_unsupported) << A->getAsString(Args);
```
Please format your code properly:
```c
} else {
  D.Diag(....);
}
```

pwisniewskimobica replied:
done

pwisniewskimobica committed:
Add more variants to test and modify CI scripts



Fix for CI job - get missing updates

Fix - apt instead of apt get

fix cosmetic and syntax remarks

install python-distutils for timeout handling

pwisniewskimobica commented:
I am closing this PR becase of @pawosm-arm ask that this task will be continued on ARM side.

pwisniewskimobica commented:
I am closing this PR becase of @pawosm-arm  ask that this task will be continued on ARM side.

bryanpkc commented:
@pawosm-arm Is there no plan to upstream this change now? I was going to approve this PR.

pawosm-arm commented:
@bryanpkc I'll prepare a new PR some time next week. This current solution shifts to much of a burden to flang2 and introduces elements of architectural awareness in a place which should be kept architecture independent.
